spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/AggregatePath.java

```diff
@@ -300,27 +304,114 @@ static TableInfo of(AggregatePath path) {
 				qualifierColumnType = path.getRequiredLeafProperty().getQualifierColumnType();
 			}
 
-			SqlIdentifier idColumnName = leafEntity.hasIdProperty() ? leafEntity.getIdColumn() : null;
+			ColumnInfos idColumnInfos = computeIdColumnInfos(tableOwner, leafEntity);
 
-			SqlIdentifier effectiveIdColumnName = tableOwner.isRoot() ? idColumnName : reverseColumnInfo.name();
+			return new TableInfo(qualifiedTableName, tableAlias, reverseColumnInfos, qualifierColumnInfo, qualifierColumnType,
+					idColumnInfos);
 
-			return new TableInfo(qualifiedTableName, tableAlias, reverseColumnInfo, qualifierColumnInfo, qualifierColumnType,
-					idColumnName, effectiveIdColumnName);
+		}
+
+		private static ColumnInfos computeIdColumnInfos(AggregatePath tableOwner,
+				RelationalPersistentEntity<?> leafEntity) {
 
+			ColumnInfos idColumnInfos = ColumnInfos.empty(tableOwner);
+			if (!leafEntity.hasIdProperty()) {
+				return idColumnInfos;
+			}
+
+			RelationalPersistentProperty idProperty = leafEntity.getRequiredIdProperty();
+			AggregatePath idPath = tableOwner.append(idProperty);
+
+			if (idProperty.isEntity()) {
+				ColumInfosBuilder ciBuilder = new ColumInfosBuilder(idPath);
+				idPath.getRequiredLeafEntity().doWithProperties((PropertyHandler<RelationalPersistentProperty>) p -> {
+					AggregatePath idElementPath = idPath.append(p);
+					ciBuilder.add(idElementPath, ColumnInfo.of(idElementPath));
+				});
+				return ciBuilder.build();
+			} else {
+				ColumInfosBuilder ciBuilder = new ColumInfosBuilder(idPath.getParentPath());
+				ciBuilder.add(idPath, ColumnInfo.of(idPath));
+				return ciBuilder.build();
+			}
 		}
-	}
 
-	record ColumnInfo(
+		private static ColumnInfos computeReverseColumnInfo(AggregatePath path) {
+
+			AggregatePath tableOwner = AggregatePathTraversal.getTableOwningPath(path);
+
+			if (tableOwner.isRoot()) {
+				return ColumnInfos.empty(tableOwner);
+			}
+
+			AggregatePath idDefiningParentPath = tableOwner.getIdDefiningParentPath();
+			RelationalPersistentProperty leafProperty = tableOwner.getRequiredLeafProperty();
+
+			RelationalPersistentProperty idProperty = idDefiningParentPath.getLeafEntity().getIdProperty();
+
+			if (idProperty != null) {
```

Comment:
These code paths result in the same downstream activity. The differences are single vs. milti-column and which entity (actual or embedded) is used to determine columns.

This should be unified.

Reply:
Not sure, if this is sufficient. If not I need further pointers.